Death Knight Runes and breath stuff

syrifgit opened the pull request:
# Overhaul on Frost DK for breath of sindragosa, and rune handling. 

## Key points
- All breath related aura events moved into the resource model
- Breath duration formula made more accurate
- Forecast Function
  - Dynamic timeout based on rune CDs and missing runes
  - Fewer redundant calculations
  - No excessive iterations with high haste values
  - Reduced overhead for large scale predictions (I think)
  - Aura and Channel filtering
  - Event prioritization and sorting based on time
- DK Rune Model
  - Stronger rune forecasting
    - time-to-max, time-to-next
  - Fewer redundant Calculations
Areas of uncertainty
- The model sorting .. is it unnecessary?
  - Could change to only sort on insertion/update of events

## CPU Performance

Cumulative updates, note that stage 5 is not included, I have stuck with the initial APL. All tests run using a 25 second breath window on pull

![image](https://github.com/user-attachments/assets/08dc2e7e-a08a-44fe-bc22-0acb95a9f54c)



Final version I tested while creating the PR, this is stage 4 -> PR (stage 5 omitted from table above, still Mondo's APL)


![image](https://github.com/user-attachments/assets/926d588d-5648-4618-9a06-950507fa92f2)


Final result:
`GetNextPrediction` during intensive breath window: Down from 12.3ms average to sub-6ms average
`Mean Clock Time` down from 51.2ms to 34ms
`Mean Work Time` down from 30.4ms to 24ms (increased slightly from stage 4 due to increased accuracy of rune system and breath duration on final version)

syrifgit committed:
Death Knight Runes

bjth commented:
Nit picky, but in the Frost LUA there are a lot of function () calls with spaces and some without. Basically just general formatting around brackets.

syrifgit commented:
> Nit picky, but in the Frost LUA there are a lot of function () calls with spaces and some without. Basically just general formatting around brackets.

Should roughly be 

`function()`
`function( with_args )`
`function () return simple thing on one line`

(I think)

bjth commented:
Just reset on Runes then, has a space before the () and isn't a single line return.

syrifgit committed:
spaces

syrifgit commented:
Draft mode, found something to fix in the BoS definition / meta thing

syrifgit committed:
Merge branch 'Hekili:thewarwithin' into DK-Rune-Model

syrifgit committed:
Update DeathKnightFrost.lua

syrifgit commented:
actually ready now. Just have a small apl edit to add if you like the overall change.

Hekili commented on State.lua:
```diff
@@ -1393,60 +1392,64 @@ do
         if not resource then return end
 
+        -- Initialize or wipe tables
+        events = events or {}
```
`events` and `remains` should be initialized as locals up above (inside the `do` but before the `forecastResources` declaration).  We wipe and reuse the same tables here so that people don't get worried about memory usage (that the garbage collector would deal with anyway).  You shouldn't ever see these as `nil` unless you're assigning them to nil somewhere.

Hekili commented on State.lua:
```diff
@@ -1393,60 +1392,64 @@ do
         if state.class.file == "DEATHKNIGHT" and state.runes then
-            timeout = max( timeout, 0.01 + 2 * state.runes.cooldown )
+            -- Adjust timeout based on rune cooldowns and regen models for Frost DK
+            timeout = max( timeout, 0.01 + state.runes.cooldown * ( 6 - state.runes.current ) )
```
This, in theory, can put the timeout at 60 seconds with 0 haste and you somehow spent all 6 runes at once.  Forecasting beyond 15 seconds doesn't actually accomplish anything.  I think this would short circuit upon reaching max runes anyway, but realistically, I think this keeps it constrained to useful values:

`timeout = max( timeout, 0.01 + state.runes.expiry[ 6 ] - query_time )`

That is assuming `expiry` is still sorted (and resorted) such that actual runes are reordered based on when they come back up.

syrifgit replied:
Yep, they are sorted.

Hekili commented on State.lua:
```diff
@@ -1393,60 +1392,64 @@ do
         remains[ resource ] = timeout
+        r.times = r.times or {}
+        r.values = r.values or {}
+        r.forecast = r.forecast or {}
```
These are technically sane, but these tables are all created upon `RegisterResource` so this shouldn't be needed.

syrifgit replied:
Maybe all my nil table issues were a scope thing, I'll take a look at them all.

Hekili commented on State.lua:
```diff
@@ -3497,39 +3494,56 @@ function state:CombinedResourceRegen( t )
-    elseif t.current >= amount then return 0 end
+    -- Handle invalid or overflow cases.
+    if not amount or amount > t.max then return 3600 end
+    if t.current >= amount then return 0 end
```
Haha, you got me.  Good cleanup.

Hekili commented on TheWarWithin/DeathKnightUnholy.lua:
```diff
@@ -150,20 +131,17 @@ me:RegisterResource( Enum.PowerType.Runes, {
-            return t.current == 6 and 0 or max( 0, t.expiry[6] - state.query_time )
-
-
-        elseif k == "add" then
```
What was `t.add` for?  I genuinely don't know.  Is it referenced anywhere?

syrifgit replied:
Looks like a deprecated version of `gain()` to me, zero references to it in state file in the context of resources. Unholy has some weird naming conventions and stuff. For example it's the only spec that uses `me.x` instead of `spec.x`

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
-            local r = state.runes
-            val = math.floor( val )
-
+            val = floor( val )
```
Is there a `local floor = math.floor` tucked away somewhere above?  Probably.

syrifgit replied:
Do they need to be added to each class file? I seem to remember it yelling at me in LUA errors when I used math.floor, and then it stopped yelling and started working when I just used floor like the other classes.

Seems like floor = math.floor is defined more globally since all class files use it without declaring (except 2 specs which declare it for some reason?)

State.lua:669 `state.floor = math.floor`

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
     cooldown = 10,
     regen = 0,
     max = 6,
-    forecast = {},
```
Oh, you took them out of here.  I think I'd rather keep them here and revert those initializations in/near `forecastResources`.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
-            duration = duration or ( 10 * state.haste )
-
-            t.expiry[ i ] = ready and 0 or start + duration
-            t.cooldown = duration
```
Are we (re)initializing `t.cooldown` somewhere else?  It needs to get reset here because haste procs will change it.

syrifgit replied:
this one was an oopsie, restored it.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
-
         for i = 1, amount do
-            t.expiry[ 7 - i ] = 0
+            table.insert( t.expiry, 0 )
```
This is clever; not sure if table.insert is more efficient but it's probably trivial.

bjth replied:
Table insert is actually slower, function calls in LUA are generally slower than using language syntax.

`t.expiry[#t.expiry+1] = 0;`

Would be faster.

syrifgit replied:
500K iteration test of each method.

Method 1 is as written, method 2 is my method, method 3 is Mondos. Table insert is faster, yours is slightly more fast most of the time. The insertion is faster because you maintain a sequential table the whole time

![image](https://github.com/user-attachments/assets/595891bc-6595-412b-9189-f4d486de6fb6)

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
-            t.expiry[ 7 - i ] = 0
+            table.insert( t.expiry, 0 )
+        end
+        while #t.expiry > 6 do
```
You could do away with this loop by adding `t.expiry[ 7 ] = nil` to the previous loop.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
         end
 
-        state.gain( amount * 10, "runic_power" )
+        if state.this_action == "obliterate" and state.buff.exterminate.up then
```
Good.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -15,137 +15,110 @@ local spec = Hekili:NewSpecialization( 251 )
-                        slice = v
-                    else
-                        break
+                    local slice = t.forecast[ i ]
```
Honestly, we might not even need to support slicing for `runes`.

syrifgit replied:
Rainy day project?

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -155,42 +128,42 @@ spec:RegisterResource( Enum.PowerType.Runes, {
-} ) )
+}))
+
+spec:RegisterStateExpr("breath_ticks_left", function()
```
Spaces.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -155,42 +128,42 @@ spec:RegisterResource( Enum.PowerType.Runes, {
+
+        last = function()
+            local app = state.buff.breath_of_sindragosa.applied
+            return app + floor( (state.query_time - app) )
```
Don't need nested parens.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -438,16 +412,59 @@ spec:RegisterAuras( {
         meta = {
-            remains = function( t )
-                if not t.up then return 0 end
-                return ( runic_power.current + ( runes.current * 10 ) ) / 17
```
Was this not accurate?  I'm curious.  At very least, there should be a `remains` metafunction here that returns your `...breath_remains` expression.  Probably an `expires = function( t ) return query_time + breath_remains end`.

(I forgot what you called your expression and I'm only scrolling in one direction.)

syrifgit replied:
The new one accounts for the fact that you can't spend 6 runes at once and involves your GCD timer and 2-rune chunks in the equation to try and get ticks. At lower haste values just spending 1-2 runes might not actually sustain you depending on tick timer.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -1425,9 +1414,9 @@ spec:RegisterAbilities( {
             gain( 5, "runic_power" )
             applyBuff( "empower_rune_weapon" )
-            for i = 5, 20, 5 do
+            --[[for i = 5, 20, 5 do
```
Just delete the commented blocks IMO.

syrifgit replied:
I usually leave them while the project isn't finalized, will take out now

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -1969,6 +1957,30 @@ spec:RegisterOptions( {
     package = "Frost DK",
 } )
 
+--[[ Estimation of whether or not random RP gains can happen, in an attempt to smooth out changing recommendations
```
?

syrifgit replied:
A way to smooth out rotation jumpiness, but it needs more iteration. May come back to it once the groundwork is pushed through.

Basically breath window has a bunch of RP thresholds, and a random +3 or +5 that can't be predicted can flip-flop a recommendation, and between the 2 recs it really doesn't matter a whole lot. The miniscule dps increase isn't worth the jarring rotation for the user.

![alt text](image.png)

syrifgit replied:
![image](https://github.com/user-attachments/assets/45687b66-5f9a-48e2-9aaf-353ab7b895d0)

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -1980,6 +1992,14 @@ spec:RegisterSetting( "bos_rp", 60, {
     width = "full"
 } )
 
+spec:RegisterStateExpr("breath_ticks_left", function()
```
Spacing.  Is this duplicated?  Only need to create it once.

syrifgit replied:
It sure is. I needed to de-declare it higher up and forgot to delete this one.

Hekili commented on TheWarWithin/DeathKnightFrost.lua:
```diff
@@ -2002,4 +2022,4 @@ spec:RegisterSetting( "ams_usage", "damage", {
 
-spec:RegisterPack( "Frost DK", 20241102, [[Hekili:S3ZFZnUTo(zjZnRx79howYX72TxsMPTx70TxVx7SPVx)VyRylhRxKL8jjVzZoz8N9dGu)GKIGIYwo7279MotJxjkqaqaqaqqYRDU(pU(QfEz(x)3Ch5EMJZi3HoVZ1zSZ1xL9Wg)RVAJ3878Uf(rK3A4))tjXPz7M9F9FJV5HWyVfiesJ3MmhE7QSSnPF7PNEBq2QT3mCE86ttdwVn0llioAEI3Ym8Fp)0RV6MTbHzVp66B039xFL32SvXjxF1vbR)baWblw4ZBTF68RVcB9RDCE9iNVD3S)(gee7M9D)(VUBww8Uzj(RJ)i8Gi)0m)f7M5nh7)0D)YUFH)HJE94rdDHp9)Xp ...
\ No newline at end of file
+spec:RegisterPack( "Frost DK", 20241102, [[Hekili:S3ZFZnUTo(zjZnRx79howYX72TxsMPTx70TxVx7SPVx)VyRylhRxKL8jjVzZoz8N9dGu)GKIGIYwo7279MotJxjkqaqaqaqqYRDU(pU(QfEz(x)3Ch5EMJZi3HoVZ1zSZ1xL9Wg)RVAJ3878Uf(rK3A4))tjXPz7M9F9FJV5HWyVfiesJ3MmhE7QSSnPF7PNEBq2QT3mCE86ttdwVn0llioAEI3Ym8Fp)0RV6MTbHzVp66B039xFL32SvXjxF1vbR)baWblw4ZBTF68RVcB9RDCE9iNVD3S)(gee7M9D)(VUBww8Uzj(RJ)i8Gi)0m)f7M5nh7)0D)YUFH)HJE94rdDHp9)Xp ...
```
If this is a changed APL, the .simc file needs to be updated as well and the second arg needs to change to the update date.

Hekili reviewed (changes requested):
Minor little changes; the missing APL is the most significant issue (if there are APL changes).

syrifgit committed:
Review notes

Implemented review notes, also added new pack string which adds my state expression to the breath entry point to mitigate against fall out of breath early

syrifgit commented:
Done, including 1 APL tweak (to implement my new expression). There may be APL tweaks to be done, but I wanted to have this reviewed/finalized before putting any more time into it, and have a more up-to-date base to work from.

Also updated the breath ticks math, Mondo pointed out an error to me the other day that I had forgot to paste in.
